Squirrel: local code intel for Java using tree-sitter

chrismwendt opened the pull request:
<p align="center">
<img width="623" alt="CleanShot 2022-03-03 at 02 29 09@2x" src="https://user-images.githubusercontent.com/1387653/156536337-a74b79ef-e575-4176-93d1-afc86cb912bb.png">
</p>

Squirrel is a technique from [CodeWing](https://codewing.dev/) for fast and precise code intelligence using [tree-sitter](https://tree-sitter.github.io/tree-sitter/). See the [RFC for Squirrel](https://docs.google.com/document/d/10YHHydQXmPODzeDDSAm_audheahdc8NjCqQPBV_wcN0/edit#) for more info.

**TODO**

- [x] Implement TJ's quick query idea to get broad coverage right away (could merge that alone, actually)
- [x] Wire it up to the GraphQL API
- [x] Update the code intel extensions to use Squirrel if available https://github.com/sourcegraph/code-intel-extensions/pull/771
- [x] Remove all traces of cross-file functionality and languages besides Java
- [x] Tune the tree-sitter queries for known locals (there are still some false positives/negatives)

## Test plan

Go tests.

tjdevries commented on cmd/symbols/squirrel/main.go:
```diff
@@ -0,0 +1,1273 @@
+func localCodeIntel(ctx context.Context, repoCommitPath types.RepoCommitPath, readFile ReadFileFunc) (*types.LocalCodeIntelPayload, error) {
+	root, contents, langSpec, err := parse(ctx, repoCommitPath, readFile)
+
+	localsPath := path.Join("nvim-treesitter", "queries", langSpec.nvimQueryDir, "locals.scm")
```
nit: You might want to cache these and use the parsed query as the thing you pass around

efritz commented on cmd/frontend/graphqlbackend/codeintel.graphql:
```diff
@@ -553,6 +553,66 @@ extend type GitBlob {
     Provides info on the level of code-intel support.
     """
     codeIntelInfo: CodeIntelSupport!
+
+    """
+    Provides code intelligence within the file.
+    """
+    localCodeIntel: JSONValue
+
+    """
+    A wrapper around syntactic hover and definition query methods.
+    """
+    symbolInfo(line: Int!, character: Int!): SymbolInfo
```
Can we add an experimental notice on these so we can be variable about the API for a stretch of time?

chrismwendt replied:
Yep, marked as experimental.

mrnugget commented on cmd/symbols/squirrel/http_handlers.go:
```diff
@@ -0,0 +1,217 @@
+		return
+	}
+
+	squirrel := NewSquirrelService(readFileFromGitserver, nil)
+	defer squirrel.Close()
+
+	// Compute the local code intel payload.
+	payload, err := squirrel.localCodeIntel(r.Context(), args, readFileFromGitserver)
```
You're passing `readFileFromGitserver` to the constructor and this method - is that by design or left-over?

mrnugget reviewed (commented):
Only left one comment, because I'm not sure what detail you want from review at this point. Since squirrel will likely be moved in the near future, I also wasn't sure what exactly to comment on or not. For example: `squirrel` package includes HTTP handlers that could live outside the package too, leaving only the service in there, making it a bit clearer what the package boundary is. I'd also separate the debug/utility functionality from the rest.

There's a PR annotation that says `go:embed` won't work because the file doesn't exist. As far as I can see, that file will be copied in the Dockerfile step. Does that mean this won't work locally?

Most importantly though, there also seem to be only 2 tests for 2000 lines of new code and no test for `squirrel` itself.

If we say that this is just prototyping/experimental and it'll get reworked soon, that's fine, but if not and we really want to make the `squirrel` package stable, then I think we should agree on what kind of review we want.

What do you think?

chrismwendt commented:
@mrnugget High-level "does this approach make sense" review is most useful because the details like how the code is organized and what the data exchange format looks like can change.

I cut cross-file from the scope of this PR so that it's local-only code intel (this was a takeaway from a [chat with Olaf and TJ](https://sourcegraph.slack.com/archives/CHXHX7XAS/p1646844926410289)).

I hard-coded the nvim-treesitter queries into Go files for portability and because I need to tweak them and add more info to the `@definition`s to know whether or not they're visible outside the immediate scope.

I added a test for `localCodeIntel` and the line count is inflated by the `language-extensions-map.json` file, so the line:test ratio is closer to 3:1000 😛 

I added logic to check for exported symbols so the client can tell if the references are "complete" or if it needs to search in other files as well.

efritz commented on cmd/frontend/graphqlbackend/codeintel.graphql:
```diff
@@ -1524,4 +1524,8 @@ enum SearchBasedSupportLevel {
     Universal-ctags is used for indexing this language.
     """
     BASIC
+    """
+    tree-sitter analysis is used for this language.
+    """
+    SYNTACTIC
```
As @olafurpg mentioned yesterday, we may want to roll this into the search-based badge for now to not explode terminology.

efritz replied:
This is more of a concern on the extension side, where we display a user badge, but thought I'd mention it here before I forgot about it 🐵

chrismwendt replied:
@Strum355 recommended I add this - I think we're bikeshedding at this point. I'll remove it.

efritz commented on enterprise/cmd/frontend/internal/codeintel/resolvers/graphql/support.go:
```diff
@@ -79,10 +80,31 @@ func NewSearchBasedCodeIntelResolver(language *string) gql.SearchBasedCodeIntelS
 	}
-	return string(unsupported)
+
+	ctagsLangToLang := map[string]string{
+		"C":          "cpp",
+		"C#":         "csharp",
+		"C++":        "cpp",
+		"Go":         "go",
+		"Java":       "java",
+		"JavaScript": "javascript",
+		"Lua":        "lua",
+		"OCaml":      "ocaml",
+		"PHP":        "php",
+		"Python":     "python",
+		"Ruby":       "ruby",
+		"Rust":       "rust",
+		"Sh":         "shell",
+	}
```
cc @Strum355 who is exposing similar information on the symbols service already. Can we unify (or plan to) here?

Strum355 replied:
Already communicated :smugcat: [`a15799c` (#32122)](https://github.com/sourcegraph/sourcegraph/pull/32122/commits/a15799c469297dfddabc9e1fe9691dc1eb8681b1)

Strum355 replied:
I think the variable name might be wrong/misleading here actually, the map is used for checking for squirrel language support

chrismwendt replied:
Removed the `SYNTACTIC` type along with this mapping.

efritz commented on cmd/symbols/squirrel/languages.go:
```diff
@@ -0,0 +1,1013 @@
+	"cpp": {
+		language:     cpp.GetLanguage(),
+		commentStyle: CommentStyle{}, // TODO
+		query: `
```
For organization, can we go-embed these from a directory and then reference the file contents by some fs abstraction?

chrismwendt replied:
They were originally embedded, then @mrnugget wondered if copying/embedding was going to work for local dev, and considering there are a few other places (like go build, Docker build, etc.) that would need to be kept up to date, I opted to hard-code them into Go so they're maximally portable. There isn't much benefit to having them in separate `.scm` files except for syntax highlighting (which only really applies to comments).

Strum355 commented:
We probably want to add honeycomb and/or grafana dashboards to this before it goes live. Any objections to doing it in this PR? If not, how do we wanna coordinate/execute that?

efritz commented on cmd/symbols/squirrel/service.go:
```diff
@@ -0,0 +1,56 @@
+// SquirrelService uses tree-sitter and the symbols service to analyze and traverse files to find
+// symbols.
+type SquirrelService struct {
+	readFile     ReadFileFunc
+	symbolSearch symbolsTypes.SearchFunc
```
Can you ping me when this PR is cleaned up of the cross-document stuff?

chrismwendt replied:
I removed the Go files related to cross-file logic, but kept some of these fields to reduce work and merge conflicts when I add it back in in a future PR. `symbolSearch` is always `nil` in this PR, and `readFile` only reads the current file. Let me know if you'd like me to remove all traces of cross-file for this PR.

mrnugget commented:
Hmm, so I've been thinking about this over night and I'm wondering: why do we want to add squirrel now?

I'm asking that question because with [RFC 619](https://docs.google.com/document/d/1AjZ_d0nJVHbV75IH3jZRkrGXhsv_AXp2kS4nrw2SAQ8/edit) @efritz is sketching out a new architecture that will finally bring LSIF Typed to the backend. We all agree that we want this, we're just now sure yet about the exact shape. That means the API for querying code intel from the browser will change (and the way with query it will also change in parts due to the new reference panel). The APIs between backend services will also likely change. Or at least how they're organized. Bottom line: we're at the start of re-organizing our architecture.

If we merge squirrel now, what do we gain? And are the drawbacks of merging it now considering that we'll re-organize our architecture in the near future? Does it even pay off to add tests, observability, etc. now, have squirrel behind a feature flag, and then immediately start rethinking where it should go?

chrismwendt commented:
@mrnugget It would be great to get this in for v3.38 (releasing next Thursday) so customers can have a much better code intel experience sooner.

I chatted with @efritz about the backend reorganization and it's my understanding that none of that blocks Squirrel (Eric correct me if I'm mistaken). How much extra burden would merging Squirrel now put on the reorganization effort?

chrismwendt commented:
I cut scope down to Java-only and local-only, ready for review 👀

efritz commented on cmd/symbols/squirrel/languages.go:
```diff
@@ -0,0 +1,87 @@
+type LangSpec struct {
+	language     *sitter.Language
+	commentStyle CommentStyle
+	query        string
```
What specifically does this query mean to return?

chrismwendt replied:
Renamed to `localsQuery` and added a docstring.

efritz commented on cmd/symbols/squirrel/util.go:
```diff
@@ -0,0 +1,243 @@
+		return nil, err
+	}
+
+	tree, err := s.parser.ParseCtx(context.Background(), nil, contents)
```
Why background ctx here?

efritz commented on cmd/symbols/squirrel/local_code_intel.go:
```diff
@@ -0,0 +1,214 @@
+					// Print a debug message if the symbol is already defined.
+					if symbol, ok := scope[symbolName]; ok && debug {
+						lines := strings.Split(string(node.Contents), "\n")
+						fmt.Printf("duplicate definition for %q in %s (skipping)\n", symbolName, repoCommitPath.Path)
+						fmt.Printf("  %4d | %s\n", symbol.Def.Row, lines[symbol.Def.Row])
+						fmt.Printf("  %4d | %s\n", node.StartPoint().Row, lines[node.StartPoint().Row])
```
Can this not be printed to stdout?

efritz replied:
Can we use log15 here if it's actually important?

chrismwendt replied:
Dropped the debugging printfs.

efritz reviewed (approved):
I didn't find anything blocking 🎉

mrnugget commented on cmd/symbols/squirrel/hover.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+// Returns the markdown hover message for the given node if it exists.
+func findHover(node Node) *string {
```
Nitpicky, sure, but how about applying some advanced nil-pointer avoidance techniques here? 😄

```suggestion
func findHover(node Node) (string, bool) {
```

chrismwendt replied:
Turns out there's always a hover, so the pointer on `*string` was unnecessary.

mrnugget commented on cmd/symbols/squirrel/hover_test.go:
```diff
@@ -0,0 +1,89 @@
+				}
+				comments = append(comments, *symbol.Hover)
+			}
+			fmt.Printf("did not find comment %q. All comments:\n", test.want)
+			for _, comment := range comments {
+				fmt.Printf("%q\n", comment)
+			}
+			t.FailNow()
```
Why not use `t.Logf` or `t.Errorf` in combination with `t.Fatalf` here?

chrismwendt replied:
What's the difference between `t.Logf` and `fmt.Printf`?

mrnugget replied:
Most succinct explanation:

<img width="690" alt="screenshot_2022-03-30_12 01 17@2x" src="https://user-images.githubusercontent.com/1185253/160805974-14ab5709-109e-4d0c-90b0-0465ca306954.png">


Longer version: if you have multiple tests running that all print with `fmt.Printf` it's really hard to track down where the output is coming from, especially if tests run concurrently.

chrismwendt replied:
I see, `t.Logf()` groups log messages with the associated test. Fixed.

mrnugget commented on cmd/symbols/squirrel/languages.go:
```diff
@@ -0,0 +1,87 @@
+// Mapping from file extension to language name.
+var extToLang = func() map[string]string {
+	m := map[string]string{}
+	for lang, exts := range langToExts {
```
Hmmm, this feels a bit hairy: global self-invoked function depends on other global self-invoked function. Can we maybe make that init behaviour explicit and add something like `squirrel.InitLangugeExtensions()` to the package and call it in `symbols`' `main`?

chrismwendt replied:
Adding `squirrel.InitLangugeExtensions()` would add an ordering dependency and anywhere it's used we'd have to remember to call it before anything else. I don't see what's hairy about this - it neatly encapsulates the initialization of a global constant.

mrnugget commented on cmd/symbols/squirrel/local_code_intel.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	// Collect defs
+	err = forEachCapture(root.LangSpec.query, *root, func(captureName string, node Node) {
```
Would it make sense to combine both `forEachCapture` calls and have a `switch` statement on `captureName` here? Or do you need to do them in order?

(I'm asking because I'm not sure whether this runs the query twice and throws away the results in between)

chrismwendt replied:
Scopes definitely need to be collected before definitions, so these must be separate calls.

chrismwendt commented on cmd/symbols/squirrel/http_handlers.go:
```diff
@@ -0,0 +1,161 @@
+			log15.Error("failed to read file from gitserver", "err", err)
+		} else {
+			prettyPrintLocalCodeIntelPayload(debugStringBuilder, *payload, string(contents))
+			fmt.Fprintln(debugStringBuilder, "✅ /localCodeIntel repo:", args.Repo, "commit:", args.Commit, "path:", args.Path)
+
+			fmt.Println(" ")
+			fmt.Println(bracket(debugStringBuilder.String()))
+			fmt.Println(" ")
```
This whole block is intended to run only when you're specifically debugging Squirrel logic.

mrnugget replied:
Ah, I missed that `Getenv("SQUIRREL_DEBUG")` up there.

chrismwendt commented on cmd/symbols/squirrel/util.go:
```diff
@@ -0,0 +1,243 @@
+)
+
+// Nominal type for the ID of a tree-sitter node.
+type Id string
```
Renamed to `NodeId`.

chrismwendt commented:
TY for reviews, I think I addressed everything. I'll do a little more manual testing then merge.

Strum355 commented on cmd/symbols/squirrel/breadcrumbs.go:
```diff
@@ -0,0 +1,147 @@
+// Prints breadcrumbs like this:
+//
+//             v 4. found
+//             v 5. goTypeDef
+//               vvv 6. goDef
```
Is this the right way around? Wouldve thought the latter would be a `goTypeDef`, going off the name

chrismwendt replied:
The content doesn't matter for this function, so I simplified the docstring to be generic.

chrismwendt commented:
The custom status page for Rockskip has been crucial for finding perf issues in ways that opentracing/Prometheus/Grafana don't support, so it would be a loss if we replaced that with anything else. None of these conflict with each other, so we can add all of them if we want.

Filed issues for tracing:

- https://github.com/sourcegraph/sourcegraph/issues/33235
- https://github.com/sourcegraph/sourcegraph/issues/33236
